src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+
+	let startTimestamp = moment(0);
+	let endTimestamp = moment(0);
+
+	let currentReading = new Reading(meterID, -1, startTimestamp, endTimestamp);
+	let prevReading = currentReading;
+	let readingOK = true;
+	for (let index = readingRepetition; index <= rows.length; ++index) {
+		// To read data where same reading is repeated. Like E-mon D-mon meters
+		if ((index - readingRepetition) % readingRepetition === 0) {
+			if (onlyEndTime) {
+				// The startTimestamp of this reading is the endTimestamp of the previous reading
+				startTimestamp = prevReading.endTimestamp;
+				endTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			else {
+				startTimestamp = moment(rows[index - readingRepetition][1], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+				endTimestamp = moment(rows[index - readingRepetition][2], ['YYYY/MM/DD HH:mm', 'MM/DD/YYYY HH:mm']);
+			}
+			if (isCumulative && isFirst(prevReading.endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing with cumulative data.';
+			}
+			if (onlyEndTime && isFirst(prevReading.endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The first reading must be dropped when dealing only with endTimestamps.';
+			}
+			if (readingOK && startTimestamp.isSameOrAfter(endTimestamp)) {
+				readingOK = false;
+				errMsg = 'The reading end time is not after the start time.';
+				if (onlyEndTime) {
+					errMsg += ' The start time came from the previous readings end time.';
+				}
+			}
+			if (readingOK) {
+				// Check that startTimestamp is not before the previous endTimestamp
+				if (onlyEndTime && endTimestamp.isSameOrBefore(prevReading.endTimestamp)) {
+					readingOK = false;
+					errMsg = 'The reading is not after the previous reading with only end time given so we must drop the reading.';
+				}
+				else if (startTimestamp.isBefore(prevReading.endTimestamp)) {
+					if (isCumulative) {
+						readingOK = false;
+						errMsg = 'The reading start time is before the previous endTime and cumulative so OED cannot use this reading.'
+					}
+					else if (!isFirst(prevReading.endTimestamp)) {
+						//Only treat this as a warning since the readings may be sent in a different order.
+						errMsg = 'The current reading startTime is not after the previous reading\'s end time. This is expected on the first reading ever.';
+					}
+				}
+			}
+			if (readingOK && Math.abs(startTimestamp.diff(prevReading.endTimestamp)) > Tgap) {
+				if (isCumulative) {
+					readingOK = false;
+					errMsg = 'The end of the previous reading is too far from the start of the next readings in cumulative data so drop the readings.';
+				}
+				else if (!isFirst(prevReading.endTimestamp)) {
+					//Only treat this as a warning since we expect a gap in the first ever reading.
+					errMsg = 'There is a gap in time between this reading and the previous reading.';
+				}
+			}
+			if (readingOK) {
+				//This reading can be used
```

Comment:
I've moved the comment to the final place when all checks have been completed. As for the empty errMsg it should be the case that an errMsg can't be empty with readingOK not false unless the errMsg is a warning. I've added an if statement to handle this warning case here. The logic now is that:

all checks pass
 if (! (errMsg === '')){
     log (errMsg);
} 
I think this should satisfy any possible warnings and in the case there is a real error it will also be logged since all errors are couple with readingOK === false which is checked at the end of given reading number.